documentation/List-of-Action-plugins.md

```diff
@@ -142,31 +143,86 @@ This will look for a user record with `username = "js"` and update it with the _
+| `counterInit`                            |                   |
+| `customFields`                           |                   |
+| `numberFormat`                           |                   |
+| `fallbackText`                           |                   |
+| `updateRecord` (default `false`)         |                   |
+| `tableName` (default `application`)      |                   |
+| `fieldName` (default `name`)             |                   |
+| `matchField` (default `"id"`)            |                   |
+| `matchValue` (default `applicationId`)   |                   |
+| `additionalData`                         |                   |
+
+This action uses the [custom_string_patterns](https://www.npmjs.com/package/custom_string_patterns), which allows the generation of random strings using Regular Expressions, but with the added functionality of also including:
+
+- incrementing, persistent counters
+- custom replacement functions
+- object data replacement.
+
+Please see the custom_string_patterns documentation for full detail, but basically the `pattern` field specified the string to be generated using:
+
+- regular expressions generate random strings that match the regex
+- `<+ddd>` represents a "counter", with each `d` specifiying a digit (will be padded with leading 0's to fill the required length). There can be any number of different counters in the system -- they are stored in the `counters` database table with a unique name and current value. Each time a value is retrieved, the number is incremented by 1.
+- `<?field>` - the `?` specifies that this is a customReplacementFunction. A simplified version is used in this Action, so any `?` strings just represent a shorthand for a data field -- the mapping of these names to actual fields is specified in `customFields` (see below)
+- `<applicationData.username>` - this just specifies the name of a property to replace into the output string. By default, `applicationData` and `outputCumulative` are available, but other data can be specified with the `additionalData` field. If the property is missing from the supplied data objects, the output string will be replaced with the value set in `fallbackText`
+
+An example:
+
+```
+{
+  pattern: "<?tempCode>-[A-Z]{3}-<+dddd>",
+  counterName: "demo",
+  counterInit: 100,
+  customFields: {
+    tempCode: "applicationData.templateCode"
+}
+```
+
+The pattern `<?tempCode>-[A-Z]{3}-<+dddd>` consists of:
+
+- `<?tempCode>` - this will be replaced with whatever data field is specified in `customFields` -- in this case `applicationData.templateCode`. Note that this is equivalent to just putting `<applicationData.templateCode>` directly in the pattern.
+- `[A-Z]{3}` a regular expression specifying 3 random upper-case characters in the A to Z range
+- `<+dddd>` will be replaced with a four-digit number, which will be taken from the current value of the counter specifed in `counterName`, in this case "demo".
```

Comment:
It's so nice that you have explained each par of the pattern used in the example, very good for future reference. I just am a little lost with why do you have that top part and then after the example you add explanation about `<?tempCode>` (one example of <?field>) and `<+ddd>` again. Isn't the case to combine both these blocks?